File: python/generalizedabbreviation.py
class Solution(object):
    
    def generateAbbreviations(self, word):
        """
        :type word: str
        :rtype: List[str]
        """
        def ga(word, pos, cur, cnt):
        
            if pos == len(word):
                if cnt > 0:
                    cur += str(cnt)
                res.append(cur)
            else:
                ga(word, pos+1, cur, cnt+1)
                ga(word, pos+1, cur + (str(cnt) if cnt>0 else '') + word[pos] , 0)
        res = []
        ga(word, 0, "", 0)
        return res
        
    

File: cpp/generalizedabbreviation.cpp
class Solution {
public:
    vector<string> generateAbbreviations(string word) {
        dfs(word, 0, "", 0);
        return res;
    }
    
    void dfs(string &word, int pos, string path, int cnt) {
        if (pos == word.length()) {
            if (cnt > 0) path += to_string(cnt);
            res.push_back(path);
            return;
        }
        dfs(word, pos+1, path, cnt+1);
        dfs(word, pos+1, path + (cnt>0? to_string(cnt) : "") + word[pos], 0);
    }
private:
    vector<string> res;
};
